src/search/algorithms.clj

```diff
@@ -46,6 +46,47 @@
                new-came-from
                new-cost-so-far)))))
 
+(defn heuristic-search [children-fn heuristic-fn start-state goal-state & {:keys [max-states] :or {max-states 1000000}}]
+  (loop [frontier (pm/priority-map start-state 0)
+         came-from {}
+         visited #{}]
+    (if (or (empty? frontier)
+            (>= (count came-from) max-states)
+            (= (first (peek frontier)) goal-state))
+      came-from
+      (let [current (first (peek frontier))
+            children (set (children-fn current))
+            unvisited-children (clojure.set/difference children visited)
+            heuristics (map (partial heuristic-fn goal-state) unvisited-children)
+            new-frontier (reduce #(assoc %1 %2 (heuristic-fn goal-state %2)) (pop frontier) unvisited-children)
+            new-came-from (reduce #(assoc %1 %2 current) came-from unvisited-children)
+            new-visited (clojure.set/union children visited)]
+        (recur new-frontier new-came-from new-visited)))))
+
+(defn heuristic-search-a-star [cost-fn children-fn heuristic-fn start-state goal-state & {:keys [max-states] :or {max-states 1000000}}]
+  (loop [frontier (pm/priority-map start-state 0)
+         came-from {}
+         visited #{}
+         cost-so-far {start-state 0}]
+    (if (or (empty? frontier)
+            (>= (count came-from) max-states)
+            (= (first (peek frontier)) goal-state))
+      [came-from cost-so-far]
+      (let [current (first (peek frontier))
+            current-cost (cost-so-far current)
+            children (set (children-fn current))
+            unvisited-children (clojure.set/difference children visited)
```

Comment:
There are a host of issues here. You've got a quasi-random mashup of my `heuristic-search` and your implementation of A*. You don't need `unvisited-children` and `heuristics`, for example, because those values are captured in `children-to-add` and `new-frontier`.